File: GeekTrust/Command.cs
using System;

namespace GeekTrust
{
    public class CommandFactory
    {
        public static Command Get(string commandLine)
        {
            var parameters = commandLine.Split(' ');
            switch (parameters[0])
            {
                case Literals.BOOK:
                    return new BookCommand(parameters[1], parameters[2], parameters[3]);
                case Literals.ADDITIONAL:
                    return new AdditionalCommand(parameters[1], parameters[2]);
                case Literals.REVENUE:
                    return new RevenueCommand();
                default:
                    break;
            }
            return new VoidCommand();
        }

    }
    public abstract class Command
    {
        public string Name { get; set; }
        abstract public string Execute();
    }
    class AdditionalCommand : Command
    {
        private string _vehicleName;
        private TimeSpan _additionalTime;
        public AdditionalCommand(string vehicleName, string additionalTime)
        {
            _vehicleName = vehicleName;
            _additionalTime = TimeSpan.Parse(additionalTime);
        }
        public override string Execute()
        {
            try
            {
                RaceTrackManager.AdditionalTime(_vehicleName, _additionalTime);
                return Literals.SUCCESS;
            }
            catch (Exception ex)
            {
                return ex.Message;
            }
        }
    }

    class RevenueCommand : Command
    {
        public RevenueCommand()
        {
        }
        public override string Execute()
        {
            try
            {
                return RaceTrackManager.GetRevenue();
            }
            catch (Exception ex)
            {
                return ex.Message;
            }
        }
    }
    class VoidCommand : Command
    {
        public VoidCommand()
        {
            Name = "";
        }
        public override string Execute()
        {
            return "";
        }
    }
    class BookCommand : Command
    {
        private Vehicle _vehicle;
        private string _vehicleType;
        public BookCommand(string vehicleType, string vehicleName, string entryTimeStr)
        {
            Name = Literals.BOOK;
            _vehicleType = vehicleType;
            var entryTime = TimeSpan.Parse(entryTimeStr);
            _vehicle = VehicleFactory.Get(vehicleType, vehicleName, entryTime);
            //_vehicle.Print();

        }
        public override string Execute()
        {
            try
            {
                RaceTrackManager.Add(_vehicle);
                return Literals.SUCCESS;
            }
            catch (Exception ex)
            {
                return ex.Message;
            }
        }
    }
}

File: python/src/command.py
from datetime import timedelta
from src.literals import Literals
from src.race_track_manager import RaceTrackManager
from src.utility import Utility
from src.vehicle_types import VehicleFactory


class CommandFactory:
    def get(cmdline):
        commands = cmdline.split(" ")
        command = commands[0]
        if command == Literals.BOOK:
            return BookCommand(commands[1], commands[2], commands[3])
        elif command == Literals.ADDITIONAL:
            return AdditionalCommand(commands[1], commands[2])
        elif command == Literals.REVENUE:
            return RevenueCommand()
        #return commands[0]
        return VoidCommand()

class Command:
    name: str
    def execute(self):
        pass

class VoidCommand(Command):
    def __init__(self):
        pass
    def execute(self):
        return ""

class BookCommand(Command):

    def __init__(self, vehicleType, vehicleName, entryTimeStr):
        self.name = Literals.BOOK
        #self._vehicleType = vehicleType
        entryTime = Utility.parse_time(entryTimeStr)
        self._vehicle = VehicleFactory.get(vehicleType, vehicleName, entryTime)
    
    def execute(self):
        try:
            ##raise Exception("EX")
            RaceTrackManager.Add(self._vehicle)
            return Literals.SUCCESS
        
        except Exception as ex:
            return ex.args[0]

class AdditionalCommand(Command):
    
    _vehicleName: str
    _additionalTime: timedelta
    def __init__(self, vehicleName, additionalTime):
        self._vehicleName = vehicleName
        self._additionalTime = Utility.parse_time(additionalTime)
    
    def execute(self):
        try:
            RaceTrackManager.AdditionalTime(self._vehicleName, self._additionalTime)
            return Literals.SUCCESS
        except Exception as ex:
            return ex.args[0]
        
        
    

class RevenueCommand(Command):

    def __init__(self):
        pass
    def execute(self):
        try:
            return RaceTrackManager.GetRevenue()
        except Exception as ex:
            return ex.args[0]
            
        
    